### PR TITLE
MXFP8 training fixes for Megatron-FSDP, Torch-FSDP, MoE, FP8 Parameter initialization

### PR DESCRIPTION
**Pull request by sudhu2k**
Changes:

- _BaseDataParallel.finish_grad_sync: accept force_all_reduce kwarg
- torch_fully_sharded_data_parallel: TE 2.x _fp8_attrs flat-attr fallback; drop debug print
- arguments: gate FSDP2 fp8_param_gather warning on TE >= 2.12, not 2.0
- fp8_utils: MXFP8 align size 32 -> 128 for HipBLASLt
- llama2/llama3 train scripts: drop MXFP8 guards; auto-add --reuse-grad-buf-for-mxfp8-param-ag when MXFP8 + fp8-param-gather without FSDP
- MLA: Ensure the proj value are aligned with mxfp8 align size.

# What does this PR do?
Fix issues with MXFP8 training in Megatron-LM

Fixes: [#15425](https://github.com/ROCm/frameworks-internal/issues/15425)
[#15420](https://github.com/ROCm/frameworks-internal/issues/15420)

### CODE REVIEW

File: examples/llama/train_llama2.sh [hunk 1]
```diff
@@ -65,7 +65,7 @@ else
 fi
 
 MODEL_SIZE="${MODEL_SIZE:-70}"
-TP="${TP:-8}"
+TP="${TP:-1}"
```
Comment (sudhu2k) [inline examples/llama/train_llama2.sh:68]: By default TP=8 in llama2 script. So, each gemm is now seeing K dim shape as
11008/8 = 1376.
When the K dim is 1376, it isn't aligned (not divisible by 128) according to hipblasLT and this throws " Unable to find any suitable algorithms"

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -718,11 +718,11 @@ def validate_args(args, defaults={}):
         assert os.environ.get('CUDA_DEVICE_MAX_CONNECTIONS') != "1", \
             'FSDP always requires CUDA_DEVICE_MAX_CONNECTIONS value large than one'
 
-        if args.fp8_param_gather and is_te_min_version("2.0.0"):
+        if args.fp8_param_gather and not is_te_min_version("2.12.0.dev0"):
```
Comment (sudhu2k) [inline megatron/training/arguments.py:721]: Bug from upstream, fsdp2 fp8 param gather was introduced in 2.12

File: megatron/core/fp8_utils.py [hunk 3]
```diff
@@ -168,7 +168,8 @@ def _get_custom_recipe(quantizer_factory_python_path: str) -> Union[Fp8Recipe, F
 def get_fp8_align_size(fp8_recipe: Fp8Recipe) -> int:
     """Get the alignment size required for fp8 GEMM."""
     if fp8_recipe == Fp8Recipe.mxfp8:
-        return 32
+        # HipblasLT requires 128 aligned Tensors for MXFP8.
```
Comment (sudhu2k) [inline megatron/core/fp8_utils.py:171]: This is used by MoEs when requiring padding on number of tokens going to an expert when it's not divisible by 128. HipblasLT requires 128 padding for GEMM.

File: megatron/core/transformer/multi_latent_attention.py [hunk 4]
```diff
@@ -542,10 +544,19 @@ def __init__(
...
 
+        kv_down_proj_out_size = self.config.kv_lora_rank + self.config.qk_pos_emb_head_dim
+        self.kv_down_proj_mxfp8_padding = 0
+        if self.config.fp8 and self.config.fp8_recipe == Fp8Recipe.mxfp8:
+            align = get_fp8_align_size(Fp8Recipe.mxfp8)
+            remainder = kv_down_proj_out_size % align
+            if remainder != 0:
+                self.kv_down_proj_mxfp8_padding = align - remainder
+                kv_down_proj_out_size += self.kv_down_proj_mxfp8_padding
+
```
Comment (sudhu2k) [inline megatron/core/transformer/multi_latent_attention.py:555]: Changes in multi_latent_attention.py fixes issues where the projection dimension (kv_down_proj_out_size) isn't divisible by 128. In deepseekv3 example, the kv down projection dimension is 576 [512 + 64] -> not divisible by 128, hence needs padding here.

**Comment by gurpreet-dhami** [comment]
@sudhu2k 
I am waiting for this PR to be merged - It will fix this issue https://amd-hub.atlassian.net/browse/ROCM-24847

**Comment by sudhu2k** [comment]
@wenchenvincent Could you take a look at this PR, whenever you find time. Thanks!

File: examples/llama/train_llama3.sh [hunk 5]
```diff
@@ -345,11 +345,15 @@ if [ "$TE_FP8" -eq 1 ]; then
...
-            echo "Error: For Llama3 FP8_PARAM_GATHER and MXFP8 cannot be currently used together, unless FSDP=1"
-            exit
+        EXTRA_ARGS="$EXTRA_ARGS --fp8-param-gather"
+
+        # MXFP8 + DDP path: TE does not implement `replace_raw_data` for MXFP8Tensor,
+        # so the default `_ParamAndGradBuffer` storage swap fails. Reusing the grad
+        # buffer for the MXFP8 param all-gather sidesteps that path. The FSDP and
+        # Megatron-FSDP paths handle MXFP8 param all-gather natively in TE and do
+        # not need this workaround.
+        if [ "$TE_FP8_RECIPE" == "mxfp8" ] && [ "$FSDP" -ne 1 ] && [ "$MEGATRON_FSDP" -ne 1 ]; then
```
Comment (wenchenvincent) [inline examples/llama/train_llama3.sh:355]: Is this limitation also existing in upstream NVTE?

File: megatron/core/distributed/data_parallel_base.py [hunk 6]
```diff
@@ -46,14 +47,20 @@ def scale_gradients(self, scaling_factor: float) -> None:
...
         """
         Finishes grad sync (all-reduce or reduce-scatter) communication operations
         for all model gradients.
 
         When overlap_grad_reduce is set to True, waits for asynchronous communication
         calls to complete. When overlap_grad_reduce is set to False, calls synchronous
         communication ops.
+
+        NOTE: ``force_all_reduce`` is included in the signature to maintain API
+        compatibility with subclasses (e.g. ``DistributedDataParallel`` and
```
Comment (wenchenvincent) [inline megatron/core/distributed/data_parallel_base.py:60]: Is this a limitation in upstream Megatron that we're fixing?

File: examples/llama/train_llama3.sh (hunk 5 above)
Comment (sudhu2k) [reply examples/llama/train_llama3.sh:355]: Yes it is
https://github.com/NVIDIA/TransformerEngine/blob/720ec27ec8483ce401dd2eaa1c76d192d54bfc84/transformer_engine/pytorch/tensor/utils.py#L68-L69

File: megatron/core/distributed/data_parallel_base.py (hunk 6 above)
Comment (sudhu2k) [reply megatron/core/distributed/data_parallel_base.py:60]: Yes, that's right

**Review by wenchenvincent — APPROVED** [review]
